src/visual.ts

```diff
@@ -329,40 +322,59 @@ export class DotPlot implements IVisual {
             .range([dotsTotalHeight, DotPlot.MinAmountOfDots]);
 
-        const dataPointsGroup: DotPlotDataGroup[] = [],
-            minDots = minValue / (maxValue / maxDots),
-            additionalValue = minValue <= 1 ? -minValue + 1 : 0; // negative values scales incorrect
+        const minDots = minValue / (maxValue / maxDots);
+        const additionalValue = minValue <= 1 ? -minValue + 1 : 0; // negative values scales incorrect
 
-        const dotScale: LogScale<number, number> = d3.scaleLog()
+        const dotScale: LogScale<number, number> = scaleLog()
             .domain(DotPlot.getDomain(minValue + additionalValue, maxValue + additionalValue))
             .range(DotPlot.getDomain(minDots > maxDots ? 1 : minDots, maxDots))
             .clamp(true);
 
+        const dataPointsGroup: DotPlotDataGroup[] = DotPlot.generateDotPlotDataGroup(valueValues, dotScale, additionalValue, valueColumn, maxDots, yScale, settings, localizationManager, visualHost, categoryColumn, categories, formattedValues, labelFontSize);
+
+        return {
+            categoryColumn,
+            settings,
+            maxXAxisHeight,
+            labelFontSize,
+            dotsTotalHeight,
+            maxLabelWidth,
+            maxLabelHeight,
+            maxCategoryWidth,
+            dataGroups: dataPointsGroup,
+            categoryAxisName: categoryColumn.source.displayName,
+            categoryLabelHeight: DotPlot.DefaultCategoryLabelHeight
+        };
+    }
+
+    private static generateDotPlotDataGroup(valueValues: number[], dotScale: LogScale<number, number, never>, additionalValue: number, valueColumn: powerbi.DataViewValueColumn, maxDots: number, yScale: LinearScale<number, number, never>, settings: DotPlotSettingsModel, localizationManager: ILocalizationManager, visualHost: IVisualHost, categoryColumn: powerbi.DataViewCategoryColumn, categories: DotPlotChartCategory[], formattedValues: string[], labelFontSize: number) {
+        const dataPointsGroup: DotPlotDataGroup[] = [];
         for (let index: number = 0, length: number = valueValues.length; index < length; index++) {
-            const value: number = valueValues[index],
-                scaledValue: number = dotScale(value + additionalValue),
-                dataPoints: DotPlotDataPoint[] = [];
+            const value: number = valueValues[index];
+            const scaledValue: number = dotScale(value + additionalValue);
+            const dataPoints: DotPlotDataPoint[] = [];
+            const hasHighlight: boolean = valueColumn.highlights && valueColumn.highlights[index] != null;
 
             for (let level: number = 0; level < scaledValue && maxDots > DotPlot.MinAmountOfDots; level++) {
                 dataPoints.push({
                     y: yScale(level),
                     tooltipInfo: DotPlot.getTooltipData(value
-                        .toFixed(settings.labels.labelPrecision)
+                        .toFixed(settings.labels.labelPrecision.value.valueOf())
```

Comment:
please remove `.valueOf()` usage in all places, where it is not really required.

Reply:
Done.